### PR TITLE
Jetty 9.4.x 1931 rollover

### PR DESCRIPTION
**Pull request by gregw**
For #1931 
Added a protected method that is called whenever a log file is rolled over.
Support a date format of "" so that a rollover file may have the same name and a backup file is created.

### CODE REVIEW

**Comment by gregw** [comment]
While writing this code, I have a concern about how the rollover stream is switched.  It is currently done with code:
```java
                OutputStream oldOut=out;
                out=new FileOutputStream(file.toString(),_append);
                if (oldOut!=null)
                    oldOut.close();
```
While this is done within a synchronised block, the protected field `out` is not volatile nor is it usage in `FilterOutputStream` synchronised.    Thus there is no memory barrier between the usage of `out` and it being closed by the rollover.  While I have not heard reports of problems,  I'm wondering if we should base this class on a VolatileFilterOutputStream class which has the `out` field as a volatile?

**Comment by gregw** [comment]
I've pushed a version that uses a volatile to hold the out.   While that is probably better than before, I think that without proper mutex on the write/flush methods, it is still possible for one thread to be writing to the outputstream while another closes it.    So either we need to delay the close (perhaps scheduling it?) or have the cost of mutex on every log write.    Note that the main user of this class is Jetty logging, which already has a mutex within the wrapping `PrintStream`, but that cannot be seen by the rollover logic to exclude the rollover/close.

**Comment by gregw** [comment]
I've talked myself into a version that uses mutex to avoid write/close race.   This race has probably existed for 20 years without a problem, but once seen has to be fixed!

File: jetty-util/src/main/java/org/eclipse/jetty/util/RolloverFileOutputStream.java [hunk 1]
```diff
@@ -335,35 +334,53 @@ void removeOldFiles(ZonedDateTime now)
...
             finally
             {
-                out=null;
+                _out=null;
                 _file=null;
             }
-    
+        }
+
+        synchronized(RolloverFileOutputStream.class)
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/RolloverFileOutputStream.java]: Is synchronized(RolloverFileOutputStream.class) here and above you made it synchronized(this), is that right?

Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/RolloverFileOutputStream.java]: The class muted is used only when accessing the shared static timer, which is done by the cancel call

**Comment by omehegan** [comment]
Can someone tell me what the timeline would be for the next 9.4 release, which I presume will include this fix? I tried to figure out if there is a formal release schedule but was unable to find anything documented. Thanks!

**Comment by sbordet** [comment]
@omehegan Jetty 9.4.8 has been staged and it's undergoing testing, so expect 1-2 days.

**Comment by omehegan** [comment]
@sbordet thanks!

**Comment by joakime** [comment]
@omehegan if you want to participate in the staged release testing ...
http://dev.eclipse.org/mhonarc/lists/jetty-dev/msg03122.html

**Pull request merged**